### ISSUE TITLE
Refactor and clean up e2e framework utils

### ISSUE THREAD

**Issue by andrewsykim** [kubernetes/kubernetes#76206]
As an initial step for https://github.com/kubernetes/kubernetes/issues/75601, we agreed to refactor & clean up miscellaneous utils code under `test/e2e/framework/*`. This is a tracking issue to go through each utils file and do the following:
* clean up any code that is unused or is safe to remove 
* clean up any code that can be consolidated with existing libraries
* should be moved into it's own logical package under `test/e2e/framework/` (or another location where appropriate)

Checklist: 
- [x] test/e2e/framework/authorizer_util.go (@johnSchnake, @neolit123)
- [x] test/e2e/framework/deployment_util.go (@alejandrox1)
- [x] test/e2e/framework/exec_util.go (@SataQiu)
- [x] test/e2e/framework/gpu_util.go (@jiatongw)
- [x] test/e2e/framework/jobs_util.go (@draveness)
- [ ] test/e2e/framework/metrics_util.go (@draveness)
- [ ] test/e2e/framework/networking_utils.go (@SataQiu)
- [ ] test/e2e/framework/nodes_util.go (@draveness)
- [ ] test/e2e/framework/perf_util.go (@WanLinghao)
- [ ] test/e2e/framework/psp_util.go (@WanLinghao)
- [ ] test/e2e/framework/pv_util.go (@WanLinghao)
- [x] test/e2e/framework/rc_util.go (@draveness)
- [x] test/e2e/framework/rs_util.go (@draveness)
- [ ] test/e2e/framework/service_util.go (@jiatongw)
- [ ] test/e2e/framework/statefulset_utils.go (@goodluckbot)
- [ ] test/e2e/framework/upgrade_util.go (@jiatongw)
- [ ] test/e2e/framework/util.go see https://github.com/kubernetes/kubernetes/issues/77095
- [x] test/e2e/framework/volume_util.go (@SataQiu)

I think there are some files under `test/e2e/framework/` not marked as a "utils" file that might be worth going over in the future, we can expand the checklist in the future to include those as well. 

/sig testing
/priority important-longterm
/assign @timothysc @neolit123 @stevesloka @johnSchnake 
/cc @alejandrox1 @jiatongw

**Comment by andrewsykim** [comment]
Feel free to comment which files you'd like to take and I'll assign it you in the issue description

**Comment by draveness** [comment]
I want to take the files below. :)

```
test/e2e/framework/jobs_util.go
test/e2e/framework/rc_util.go
test/e2e/framework/rs_util.go
```

**Comment by alejandrox1** [comment]
I'd like to take`test/e2e/framework/deployment_util.go`

**Comment by stevesloka** [comment]
I can take: 

```
test/e2e/framework/nodes_util.go
```

**Comment by jiatongw** [comment]
I can take `gpu_util.go` and `upgrade_util.go`

**Comment by jiatongw** [comment]
@andrewsykim how about we create a `utils/` folder under `framework/` folder,  and move most of the files end with `_util.go` into `utils` folder?

**Comment by goodluckbot** [comment]
I'd like to take
```
test/e2e/framework/statefulset_utils.go
```

**Comment by draveness** [comment]
I checked the job_util and rc_util files and found out there are many generic functions like:

```go
// rc_util.go
func trimDockerRegistry(imagename string) string {
	imagename = strings.Replace(imagename, "docker.io/", "", 1)
	return strings.Replace(imagename, "library/", "", 1)
}

// job_util.go
func newBool(val bool) *bool {
	p := new(bool)
	*p = val
	return p
}
```

At first, I thought I could put these methods into `util.go`. However, it's a file with more than 5,000 lines of code. My question is do we need a new place to put these generic functions or just move these functions into `util.go`. And I do think it's going to be hard to maintain such a huge file in the future. 🤣

**Comment by WanLinghao** [comment]
I would like to take:
 test/e2e/framework/perf_util.go
test/e2e/framework/psp_util.go
test/e2e/framework/pv_util.go

**Comment by andrewsykim** [comment]
> @andrewsykim how about we create a utils/ folder under framework/ folder,  and move most of the files end with _util.go into utils folder? The current _util.go files just leave there and not good organized, however, move those into utils/ folder also needs to change a lot bit in many files in test/ since it changes the import path


@jiatongw IMO I don't think `test/e2e/framework/utils` directory is necessary, but let's see how the first few PRs go and see if adding it can be beneficial.

**Comment by andrewsykim** [comment]
@draveness for unexported functions I would recommend checking the callers of the function first, we may be able to fold those into fewer functions and from there it may be easier to decide where that code should live.

**Comment by andrewsykim** [comment]
fwiw, I think functions like `newBool` are okay to duplicate

**Comment by andrewsykim** [comment]
Seems like we need to coordinate on what package we want to move some of these util files. Here are a few options I see going forward:
1) `test/e2e/framework/<pkg>` 
2) `test/e2e/framework/util/<pkg>`
3) `test/utils/<pkg>`

I'm in favor of 1), @johnSchnake @stevesloka @timothysc @neolit123 @alejandrox1 thoughts?

**Comment by johnSchnake** [comment]
On my PR I chose (1), partially because I had seen that used in another similar PR.

In my mind, the framework code is itself is "stuff to help you run your tests" in one form or another, so `framework/util/<pkg>` seems a slightly redundant.

Hadn't really thought about (3), pulling it all the way out of framework and e2e. Just peeking at the code the `e2e` stuff is more code/test focused which would include most of the stuff we are looking at. (The `util` pkg seemed to have more diverse, high level items like the test harness, junit, etc. There was one CRD util which seemed to be more e2e-ish though)

Loosely held belief at the moment, but those are my 2 cents.

**Comment by andrewsykim** [comment]
> On my PR I chose (1), partially because I had seen that used in another similar PR.
>
> In my mind, the framework code is itself is "stuff to help you run your tests" in one form or another, so framework/util/<pkg> seems a slightly redundant.

Agreed, let's stick with 1) for now unless others think differently :)

**Comment by SataQiu** [comment]
I would like to take:
test/e2e/framework/exec_util.go

**Comment by WanLinghao** [comment]
When I try refactor `pv_util.go`, I found that it includes too many functions and many of them has none relationship with pv/pvc at all. So I decide to split those functions into different files. I create two new directories to put them: `framework/pod  framework/pvpvc`.
But now I am in trouble with package loop import.   I think we need move util.go file to its own sub directory(this file is insane with 5000 lines) to solve this. @andrewsykim WDYT

**Comment by johnSchnake** [comment]
@WanLinghao I'm thinking that may be required as well.

I hadn't done it yet in my PR but if try and follow the general rule that we should only import files _farther_ down the path (rather than back up the path) then anything we want the `framework/<pkg>` files to share along with `framework` needs to be put in a separate `framework/util`

For me, it is the `IsRBACEnabled` method. It is used in a number of tests and other `*_util.go` files which will inevitably be put into their own packages. So I think the `framework/util`, although a bit redundant is the only way to avoid putting it all into one large package.

**Comment by johnSchnake** [comment]
So the `ProviderIs` is one of the many functions in that `util.go` file you were talking about @WanLinghao . It, like a lot of the other methods there, really just need access to the test context.

Since that is such a shared value which even our `framework/<pkg>` utils will need, maybe the larger refactor we need to do here is pushing that down into it's own `framework/testContext` package?

Obviously, TestContext is used _a lot_ so it is not a small refactor, but the shared resource needs to be pushed down farther into the stack.

If I can get agreement about ^ I'll take that on.

**Comment by draveness** [comment]
Just got two PRs approved, I'd like to take one more file :)

```
test/e2e/framework/metrics_util.go
```

**Comment by WanLinghao** [comment]
@johnSchnake @andrewsykim It seems that we should make decision about util.go file before push on.

**Comment by draveness** [comment]
> @johnSchnake @andrewsykim It seems that we should make decision about util.go file before push on.

I think maybe it's better to refactor the other files as many as possible, and then take a look up at the util.go, and find out how to move the functions in util.go file to other packages.

**Comment by alok87** [comment]
i would like to take:
`test/e2e/framework/metrics_util.go`

**Comment by jiatongw** [comment]
There are a lot of error handling methods, in `util.go` and are referred by many other files. Another big refactor could be extracting them to an error handling specific file

**Comment by SataQiu** [comment]
I would like to take:
test/e2e/framework/networking_utils.go

**Comment by jiatongw** [comment]
Hey guys, `service_util.go` is also hard to maintain: 1600+ lines of code and many of them have no comments..

**Comment by andrewsykim** [comment]
Seems like `util.go` is blocking a lot of the other files, I'm going to take a look at it today if folks don't mind

**Comment by jiatongw** [comment]
I can do a first step clean of `framework/service_util.go`

**Comment by andrewsykim** [comment]
@johnSchnake and I started writing down some ways to break up `util.go` here https://docs.google.com/document/d/1zvbr8iun_L2OkHvucnstFhcxsWS7wyqfRQ9UztBZWgw/edit. Please take a look, if we agree we can start breaking up this work into smaller chunks.

**Comment by draveness** [comment]
Is there anything else I could do for this issue? I have refactored files that I claimed. :)

**Comment by andrewsykim** [comment]
@draveness I think we can break up `utils.go` into smaller tasks, I'll update the issue description once we've agreed on how to do this

**Comment by WanLinghao** [comment]
@andrewrynhard  Do we have any progress here?

**Comment by andrewrynhard** [comment]
> @andrewrynhard Do we have any progress here?
@andrewsykim

**Comment by andrewsykim** [comment]
Sorry folks, been travelling this week so this fell off my radar. I opened up a separate issue for `test/e2e/framework/util.go`, please assign yourself to tasks there https://github.com/kubernetes/kubernetes/issues/77095

**Comment by oomichi** [comment]
@andrewsykim I know you are traveling this week, it is fine to response after that.
(I just don't want to forget this :-))
Current module names under `test/e2e/framework/<..>/` seem inconsistent and that lead this work makes more inconsistent module names.
The following is current module names:
- test/e2e/framework/auth/helpers.go
- test/e2e/framework/ginkgowrapper/wrapper.go
- test/e2e/framework/gpu/gpu_util.go
- test/e2e/framework/ingress/ingress_utils.go
- test/e2e/framework/podlogs/podlogs.go
- test/e2e/framework/testfiles/testfiles.go
- test/e2e/framework/timer/timer.go
- test/e2e/framework/volume/fixtures.go

Most simple and easy way could be `util.go` on every path, `<path name>_util.go` also is a nice option I feel.

**Comment by andrewsykim** [comment]
> Most simple and easy way could be util.go on every path, <path name>_util.go also is a nice option I feel.

Tried to avoid this since many of the existing `test/e2e/framework/*_util.go` files still need a refactor + clean up. Trying to move more files in there could make it harder to refactor later. What do you think @oomichi?

**Comment by WanLinghao** [comment]
@andrewsykim  Hi, I was just in vocation, I still want to take:
test/e2e/framework/perf_util.go
test/e2e/framework/psp_util.go
test/e2e/framework/pv_util.go

**Comment by draveness** [comment]
I could take `nodes_util.go` and `metrics_util.go`

**Comment by draveness** [comment]
for `test/e2e/framework/metrics_util.go`, the `metrics` directory has already existed. Is it appropriate to move codes to the directory?

**Comment by alejandrox1** [comment]
Moving this to the next milestone 😃 
/milestone v1.16

**Comment by alejandrox1** [comment]
Hi @draveness i noticed that test/e2e/framework/metrics_util.go hasn't been started yet, could I take it?

**Comment by draveness** [comment]
> Hi @draveness i noticed that test/e2e/framework/metrics_util.go hasn't been started yet, could I take it?

I was blocked by the Failf method which has just merged into master, please feel free to take it.

**Comment by alejandrox1** [comment]
> > Hi @draveness i noticed that test/e2e/framework/metrics_util.go hasn't been started yet, could I take it?
> 
> I was blocked by the Failf method which has just merged into master, please feel free to take it.

Thank you. I will start working on metrics_util.go today

**Comment by jiatongw** [comment]
@alejandrox1 
Updated: we can mark `upgrade_util.go` as done (it was done month ago) I am now start cleaning up the reset of e2e/framework/node package

**Comment by alejandrox1** [comment]
Hi @WanLinghao can I help out with test/e2e/framework/pv_util.go ?

**Comment by WanLinghao** [comment]
@alejandrox1 OK

**Comment by Pothulapati** [comment]
Hello! I'm part of the bug triage team for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th (about 1.5 weeks from now), which means that there should be a PR ready (and merged) until then.

Do we still target this issue to be fixed for 1.16?

**Comment by neolit123** [comment]
hi @Pothulapati the work here will be ongoing for multiple releases.
we should update the milestone to 1.17 at some point.

**Comment by alejandrox1** [comment]
bumping to the next milestone for now but we need to reevaluate this issue.
/milestone v1.18

**Comment by ttousai** [comment]
***Correction:*** Code freeze is on Thursday, November 14.

Sorry for the mix up.

**Comment by smourapina** [comment]
Hello @alejandrox1! This is just a friendly reminder from the Bug Triage team: are we still targeting this issue for milestone 1.18?

**Comment by neolit123** [comment]
thanks for the heads up @smourapina 
moving to this to 1.19.

/milestone v1.19

**Comment by andrewsykim** [comment]
Lowering the priority for this one. For v1.19 we want to prioritize https://github.com/kubernetes/kubernetes/issues/74352.

/priority backlog

**Comment by alejandrox1** [comment]
/milestone clear

Will cleanup this thing as well to better direct what has yet to be done

**Comment by spiffxp** [comment]
@oomichi is this still relevant? I don't have the bandwidth to shepherd this as help-wanted and am considering /remove-help
